extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs

```diff
@@ -0,0 +1,221 @@
+        }
+
+        // This will be for input bindings only.
+        // a) If users bind to just a POCO, they need to provide the `Id` and `PartitionKey`
+        //     attributes so that we know which document to pull
+        // b) If they bind to IList<POCO>, we should be able to just pull every document
+        //    in the container, unless they specify the the SqlQuery attribute, in which case
+        //    we need to filter on that.
+        private async Task<object> CreatePOCOFromReference<T>(CosmosDBInputAttribute cosmosAttribute)
+        {
+            var container = CreateCosmosClient<Container>(cosmosAttribute) as Container;
+
+            if (container is null)
+            {
+                // use proper exception type or handle
+                throw new InvalidOperationException("Houston, we have a problem");
+            }
+
+            if (cosmosAttribute.Id is not null && cosmosAttribute.PartitionKey is not null)
+            {
+                ItemResponse<T> item = await container.ReadItemAsync<T>(cosmosAttribute.Id, new PartitionKey(cosmosAttribute.PartitionKey));
+
+                if (item is null || item?.StatusCode is not System.Net.HttpStatusCode.OK)
+                {
+                    throw new InvalidOperationException("Unable to retrieve document with ID {cosmosAttribute.Id} and PartitionKey {cosmosAttribute.PartitionKey}");
+                }
+
+                return item.Resource;
+            }
+
+            if (cosmosAttribute.SqlQuery is not null)
+            {
+                using (var iterator = container.GetItemQueryIterator<T>(cosmosAttribute.SqlQuery))
+                {
+                    return await ExtractCosmosDocuments(iterator);
+                }
+            }
+
+            var allContainerDocuments = container.GetItemQueryIterator<T>();
+            using (var iterator = container.GetItemQueryIterator<T>())
+            {
+                return await ExtractCosmosDocuments(iterator);
+            }
+        }
+
+        private async Task<IList<T>> ExtractCosmosDocuments<T>(FeedIterator<T> iterator)
+        {
+            var documentList = new List<T>();
+            while (iterator.HasMoreResults)
+            {
+                FeedResponse<T> response = await iterator.ReadNextAsync();
+                IEnumerable<T> documents = response.Resource;
+                documentList.AddRange(documents);
+            }
+            return documentList;
+        }
+
+        private object CreateCosmosClient<T>(CosmosDBInputAttribute cosmosAttribute)
+        {
+            if (cosmosAttribute is null)
+            {
+                // What do?
+                throw new InvalidOperationException("Cosmos attribute cannot be null");
+            }
+
+            CosmosClient cosmosClient;
+            var connectionString = Environment.GetEnvironmentVariable(cosmosAttribute.Connection);
```

Comment:
Just to make sure - are you still planning on switching to `IConfiguration` for this? Looking at your `Constants.cs` file, all your properties are `AzureWebJobsStorage__*`, so `IOptions<T>` would actually be a perfect fit!

``` CSharp
internal class CosmosBindingOptions
{
    public string AccountName { get; set; }
    
    // other properties also
}

// in extension startup:
applicationBuilder.Services.AddOptions<CosmosBindingOptions>().BindConfiguration("AzureWebJobsStorage");

// inject IOptions into converter:

public CosmosDbConverter(IOptions<CosmosBindingOptions> options)
{
}
```

Reply:
Yup that's the plan! Will work on that next after making sure managed ID is working as expected

Reply:
@jviau  I went with a slightly different approach here, still using IConfiguration, thoughts?

Reply:
I would recommend using `IOptions` here, as it is more flexible that `IConfiguration`. Maybe even make the options class public. This will be beneficial for two reasons:

1. Unit testing! You now only need to create an options object, no need to set up `IConfiguration`.
2. Customers can configure the options object in a variety of ways, not just via `IConfiguration`.

My opinion is that you should only very sparingly consume `IConfiguration` directly in code. In most cases you want to use options pattern - this separates the concern between how you consume config/settings and where/how those settings are built/populated.

Reply:
An additional benefit, is you can even move the `TokenCredential` resolution from `AzureComponentFactory` into the options configuration, which simplifies your cosmos DB factory, as it will have a built `TokenCredential` provided to it.

Reply:
I'll be honest, I'm not exactly sure how this works or what it is supposed to look like but I will make an attempt! I'd love to setup time for us to chat about this so I can learn more!